src/NuGet.Indexing/IndexDirectoryProvider.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Lucene.Net.Store;
+using Lucene.Net.Store.Azure;
+using Microsoft.Extensions.Logging;
+using Microsoft.WindowsAzure.Storage;
+using NuGet.Services.Configuration;
+using FrameworkLogger = Microsoft.Extensions.Logging.ILogger;
+
+namespace NuGet.Indexing
+{
+    public class IndexDirectoryProvider : IIndexDirectoryProvider
+    {
+        private readonly FrameworkLogger _logger;
+
+        private readonly ISettingsProvider _settings;
+
+        private Directory _directory;
+        private string _indexContainerName;
+        private AzureDirectorySynchronizer _synchronizer;
+
+        public static async Task<IndexDirectoryProvider> Create(ISettingsProvider settings, FrameworkLogger logger)
+        {
+            var indexSynchronizer = new IndexDirectoryProvider(settings, logger);
+            await indexSynchronizer.Reload();
+            return indexSynchronizer;
+        }
+
+        protected IndexDirectoryProvider(ISettingsProvider settings, FrameworkLogger logger)
+        {
+            _logger = logger;
+            _settings = settings;
+        }
+
+        public Directory GetDirectory()
+        {
+            return _directory;
+        }
+
+        public string GetIndexContainerName()
+        {
+            return _indexContainerName;
+        }
+
+        public AzureDirectorySynchronizer GetSynchronizer()
+        {
+            return _synchronizer;
+        }
+
+        public async Task<bool> Reload()
+        {
+            // If we have a directory and the index container has not changed, we don't need to reload.
+            // We don't want to reload the index unless necessary.
+            var newIndexContainerName = await _settings.GetOrDefault(IndexingSettings.IndexContainer, IndexingSettings.IndexContainerDefault);
+            if (_directory != null && newIndexContainerName == _indexContainerName)
+            {
+                return false;
+            }
+
+            _indexContainerName = newIndexContainerName;
+            _logger.LogInformation("StorageLoader index container: {IndexContainerName}", _indexContainerName);
+            
+            var storageAccount = CloudStorageAccount.Parse(await _settings.GetOrThrow<string>(IndexingSettings.StoragePrimary));
```

Comment:
Does the setting for IndexingSettings.StoragePrimary contain a full connection string? If it's just the storage key, CloudStorageAccount.Parse I don't think can create a valid connection. Can anybody confirm?